src/components/AddApplication/AddApplication.js

```diff
@@ -0,0 +1,138 @@
+import React, { useState } from 'react';
+import PropTypes from 'prop-types';
+import { withRouter } from 'react-router-dom';
+import { connect } from 'react-redux';
+import { bindActionCreators } from 'redux';
+import { FormattedMessage, injectIntl } from 'react-intl';
+import { Wizard, Text, TextVariants, TextContent } from '@patternfly/react-core';
+
+import { loadEntities } from '../../redux/actions/providers';
+import SourcesFormRenderer from '../../Utilities/SourcesFormRenderer';
+import createSchema from './AddApplicationSchema';
+import LoadingStep from '../steps/LoadingStep';
+import ErroredStep from '../steps/ErroredStep';
+import FinishedStep from '../steps/FinishedStep';
+
+import { doCreateApplication } from '../../api/entities';
+
+const AddApplication = (
+    { source, appTypes, history, loadEntities, intl, appTypesLoaded, sourceTypesLoaded }
+) => {
+    const [state, setState]  = useState('wizard');
+    const [error, setError]  = useState('');
+    const [values, setValues] = useState({});
+
+    if (!source || !appTypesLoaded || !sourceTypesLoaded) {
+        return null;
+    }
+
+    const appIds = source.applications.reduce((acc, app) => [...acc, app.application_type_id], []);
+    const filteredAppTypes = appTypes.filter((type) => !appIds.includes(type.id));
+    const schema = createSchema(filteredAppTypes.map((type) => ({ value: type.id, label: type.display_name })), intl);
+
+    const onSubmit = ({ application }) => {
+        setState('loading');
+        return doCreateApplication(source.id, application).then(() => {
+            setState('finished');
+            loadEntities();
+        })
+        .catch(({ data: { errors: [{ detail }] } }) => {
+            setState('errored');
```

Comment:
This will invoke 3 re-renders. It is recommended to use `useReducer` instead of `useState` to optimize cases when you are calling more than one stet update at once.